### DISCUSSION TITLE
What's the future way to use DISTINCT ON with Postgresql dialect?

### DISCUSSION THREAD

**Discussion by simsicon** [sqlalchemy/sqlalchemy#7091] [Usage Questions] [answered]
Hi I have ran into a SADeprecationWarning says

> DISTINCT ON is currently supported only by the PostgreSQL dialect.  Use of DISTINCT ON for other backends is currently silently ignored, however this usage is deprecated, and will raise CompileError in a future release for all backends that do not support this syntax.

Also found in page: https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.distinct

But what's the future and compatible way to use the dialect in future version? Thanks.

**Comment by CaselIT** [comment, marked as answer]
I'm not sure about the question.

The warning just means that distinct on works only on postgresql and it will raise an error if used on other db.

Could you rephrase?

**Reply by CaselIT** [reply]
You mean that you are getting this error while using postgresql?

Can you provide an example of how you are getting the warning?

**Reply by simsicon** [reply]
> I'm not sure about the question.
> 
> The warning just means that distinct on works only on postgresql and it will raise an error if used on other db.
> 
> Could you rephrase?

Thank you for your explaining, I misunderstood the warning, I thought this distinct on expr usage is deprecating even for postgresql.

But I do get this SADeprecationWarning while I am using postgresql+psycopg2, so I think maybe the warning is redundant?

I have crafted the reproduce code here https://github.com/simsicon/sa_test, please run it with pytest to see the warning, I also attached a screenshot.

**Reply by CaselIT** [reply]
The issue here is of pandas. When using `read_sql` with a statement pandas tries to check if a table exists using the input statement that gets converted to string. Use `read_sql_query` to avoid it.
